pkg/resource/training_job/custom_set_update_input.go

```diff
@@ -0,0 +1,130 @@
+	if delta.DifferentAt("Spec.ProfilerConfig") && delta.DifferentAt("Spec.ProfilerRuleConfigurations") {
+		err := handleProfilerRuleConfig(desired, latest, input)
+		return err
+	}
+	if delta.DifferentAt("Spec.ProfilerConfig") && !delta.DifferentAt("Spec.ProfilerRuleConfigurations") {
+		input.SetProfilerRuleConfigurations(nil)
+		return nil
+	}
+	if delta.DifferentAt("Spec.ProfilerRuleConfigurations") && !delta.DifferentAt("Spec.ProfilerConfig") {
+		err := handleProfilerRuleConfig(desired, latest, input)
+		input.SetProfilerConfig(nil) // SM still assumes the profiler config is the same.
+		return err
+	}
+	return nil
+}
+
+// handleProfilerRuleConfig sets the input of the ProfilerRuleConfiguration so that
+// it is compatible with the sagemaker API.
+// Update training job is post operation wrt to the profiler parameters.
+// Because of this only NEW rules can be specified.
+// In this function we check to see if any new profiler configurstions have been added.
+func handleProfilerRuleConfig(desired *resource, latest *resource, input *svcsdk.UpdateTrainingJobInput) error {
+	profilerRuleDesired := desired.ko.Spec.ProfilerRuleConfigurations
+	profilerRuleLatest := latest.ko.Spec.ProfilerRuleConfigurations
+
+	if ackcompare.IsNil(profilerRuleDesired) {
+		return ackerr.NewTerminalError(errors.New("cannot remove a profiler rule."))
+	}
+	if ackcompare.IsNil(profilerRuleLatest) {
+		return nil
+	}
+	if len(profilerRuleDesired) < len(profilerRuleLatest) {
+		return ackerr.NewTerminalError(errors.New("cannot remove a profiler rule."))
+	}
+
+	ruleMap := map[string]int{}
+	profilerRuleInput := []*svcsdk.ProfilerRuleConfiguration{}
+	for _, rule := range profilerRuleLatest {
+		if ackcompare.IsNotNil(rule) && ackcompare.IsNotNil(rule.RuleConfigurationName) {
+			ruleMap[*rule.RuleConfigurationName] = 1
+		}
+	}
+	for _, rule := range profilerRuleDesired {
+		if ackcompare.IsNotNil(rule) && ackcompare.IsNotNil(rule.RuleConfigurationName) {
+			_, present := ruleMap[*rule.RuleConfigurationName]
+			if !present {
+				profilerRuleInput = append(profilerRuleInput, convertProfileRuleType(rule))
+			}
+		}
+	}
+	input.SetProfilerRuleConfigurations(profilerRuleInput)
+	return nil
+}
+
+// handleProfilerRemoval sets the input parameters to disable the profiler.
+func handleProfilerRemoval(input *svcsdk.UpdateTrainingJobInput) {
+	input.SetProfilerRuleConfigurations(nil)
+	profilerConfig := svcsdk.ProfilerConfigForUpdate{}
+	profilerConfig.SetDisableProfiler(true)
+	input.SetProfilerConfig(&profilerConfig)
+}
+
+// convertProfileRuleType converts the kubernetes object ProfilerRuleConfiguration into
+// a type that is compatible with the AWS API.
+// Sagemaker and kubernetes types are not the same so the input has to be reconstructed.
+func convertProfileRuleType(rule *smv1alpha.ProfilerRuleConfiguration) *svcsdk.ProfilerRuleConfiguration {
+	smRule := &svcsdk.ProfilerRuleConfiguration{}
```

Comment:
rename to rule, all of the code is related to SM

Reply:
Yes but smRule is a sagemaker api type and rule is a kubernetes object type.